### PR TITLE
feat(ThumbnailGridBlock): refactor storing

### PR DESCRIPTION
**Pull request by fulopdaniel**
- Placeholder item has been removed, instead an actual empty item is added every time it's needed. (condition: last item is not empty)

- Uploading is happening in the child (Item.tsx) component instead of the parent (ThumbnailGridBlock.tsx) component. The child notifies the parent once the uploaded finishes and the appropriate handlers are called.

- Parent keeps track of the upload loading states of the children. Essentially to show the loading circle or not.

- It is possible to add several images at once.

### CODE REVIEW

File: packages/thumbnail-grid-block/src/ThumbnailGridBlock.tsx [hunk 1]
```diff
@@ -1,138 +1,102 @@
...
+            id: generateRandomId(),
+        }));
+        addItems(newItemsToAdd);
+
+        return newItemsToAdd;
+    };
+
+    const onAssetsSelected = async (selectedAssets: Asset[], itemId: string) => {
+        const newItemsToAdd = addNewItemsIfNeeded(selectedAssets);
+        const _uploadingItemIds = {
```
Comment (ragi96) [inline packages/thumbnail-grid-block/src/ThumbnailGridBlock.tsx:39]: Improve naming of that variable

**Review by ragi96 — CHANGES_REQUESTED** [review]
Uploading multiple crashes block:
Screenshot 2023-09-26 at 10 30 31

File: packages/thumbnail-grid-block/src/ThumbnailGridBlock.spec.ct.tsx [hunk 2]
```diff
@@ -54,14 +53,13 @@ describe('Thumbnail Grid Block', () => {
...
-        cy.get(ThumbnailItem).should('not.exist');
+        cy.get(ThumbnailItem).should('exist');
         cy.get(ThumbnailImage).should('not.exist');
         cy.get(ThumbnailImagePlaceholder).should('exist');
         cy.get(ThumbnailImageWrapper).should('have.class', 'tw-place-self-start');
-        cy.get(ThumbnailItemPlaceholder).should('have.class', 'tw-flex-row');
     });
 
-    it('should render one item if it is provided', () => {
+    it('should render one item if it is provided and automatically add an empty one', () => {
```
Comment (ragi96) [inline packages/thumbnail-grid-block/src/ThumbnailGridBlock.spec.ct.tsx:62]: Can you add a check that the second one is an empty one?

File: packages/thumbnail-grid-block/src/ThumbnailGridBlock.tsx [hunk 3]
```diff
@@ -1,138 +1,101 @@
...
+        },
+        [itemsState, setItemsState],
+    );
 
-    const addItem = (id: string, type: keyof Thumbnail, value: string, altText?: string) => {
-        setItemsState((old) => [...old, { id, [type]: value, altText }]); // cannot use saveItems here, as it overwrites the array when adding multiple image
+    const updateItems = (updatedItems: Thumbnail[]) => {
+        const newItemsState = itemsState.map((item) => {
+            const updatedItem = updatedItems.find((i) => i.id === item.id);
+            return updatedItem || item;
```
Comment (ragi96) [inline packages/thumbnail-grid-block/src/ThumbnailGridBlock.tsx:85]:
```suggestion
            return updatedItem ?? item;
```

File: packages/thumbnail-grid-block/src/components/Thumbnail/Item.tsx [hunk 4]
```diff
@@ -27,28 +35,53 @@ export const Item = forwardRef<HTMLDivElement, ThumbnailItemProps>(
...
+            });
         };
-        const onAssetChooserClick = () => openAssetChooser(id);
+
+        useEffect(() => {
+            if (selectedFiles) {
+                onFilesSelected(selectedFiles, id);
+                uploadFile(selectedFiles);
+            }
+        }, [selectedFiles]);
```
Comment (ragi96) [inline packages/thumbnail-grid-block/src/components/Thumbnail/Item.tsx:78]:
```suggestion
        }, [selectedFiles, onFilesSelected, uploadFile, id]);
```

File: packages/thumbnail-grid-block/src/components/Thumbnail/Item.tsx [hunk 5]
```diff
@@ -27,28 +35,53 @@ export const Item = forwardRef<HTMLDivElement, ThumbnailItemProps>(
...
+
+        useEffect(() => {
+            if (doneAll) {
+                onFilesUploaded(uploadResults, id);
+            }
+        }, [doneAll, uploadResults]);
```
Comment (ragi96) [inline packages/thumbnail-grid-block/src/components/Thumbnail/Item.tsx:84]:
```suggestion
        }, [doneAll, uploadResults, onFilesUploaded, id]);
```

File: packages/thumbnail-grid-block/src/ThumbnailGridBlock.tsx [hunk 6]
```diff
@@ -1,138 +1,101 @@
...
-    useBlockSettings,
-    useEditorState,
-    useFileInput,
-} from '@frontify/app-bridge';
+import { Asset, useBlockAssets, useBlockSettings, useEditorState } from '@frontify/app-bridge';
 
 import { BlockProps, gutterSpacingStyleMap, useDndSensors } from '@frontify/guideline-blocks-settings';
 import { generateRandomId } from '@frontify/fondue';
 
 import { Settings, Thumbnail } from './types';
```
Comment (ragi96) [inline packages/thumbnail-grid-block/src/ThumbnailGridBlock.tsx:15]:
```suggestion
import types { Settings, Thumbnail } from './types';
```

File: packages/thumbnail-grid-block/src/types.ts [hunk 7]
```diff
@@ -8,7 +8,6 @@ export type Thumbnail = {
     id: string;
```
Comment (ragi96) [inline packages/thumbnail-grid-block/src/types.ts:8]: all imports here should be `import types`

**Review by ragi96 — COMMENTED** [review]
Shouldn't be able to move the empty element, also the replace by asset / upload seems odd on an empty one

File: packages/thumbnail-grid-block/src/ThumbnailGridBlock.spec.ct.tsx (hunk 2 above)
Comment (fulopdaniel) [reply packages/thumbnail-grid-block/src/ThumbnailGridBlock.spec.ct.tsx:62]: Sure, done!

File: packages/thumbnail-grid-block/src/components/Thumbnail/Item.tsx (hunk 4 above)
Comment (fulopdaniel) [reply packages/thumbnail-grid-block/src/components/Thumbnail/Item.tsx:79]: Those dependencies change too often and can result in faulty functioning. To be in line with other uses of this hook I added a line to disable the warning.

File: packages/thumbnail-grid-block/src/ThumbnailGridBlock.tsx (hunk 6 above)
Comment (fulopdaniel) [reply packages/thumbnail-grid-block/src/ThumbnailGridBlock.tsx]: Fixed

File: packages/thumbnail-grid-block/src/components/Thumbnail/Item.tsx (hunk 4 above)
Comment (ragi96) [reply packages/thumbnail-grid-block/src/components/Thumbnail/Item.tsx:79]: Thought they wouldn't change so its okay.

File: packages/thumbnail-grid-block/src/components/Thumbnail/Item.tsx [hunk 8]
```diff
@@ -75,14 +110,14 @@ export const Item = forwardRef<HTMLDivElement, ThumbnailItemProps>(
                             : [],
                         [
                             {
-                                title: 'Replace with upload',
+                                title: image ? 'Replace with upload' : 'Upload image',
```
Comment (ragi96) [inline packages/thumbnail-grid-block/src/components/Thumbnail/Item.tsx:113]:
```suggestion
                                title: image ? 'Replace with upload' : 'Upload asset',
```

File: packages/thumbnail-grid-block/src/components/Thumbnail/Item.tsx [hunk 9]
```diff
@@ -75,14 +110,14 @@ export const Item = forwardRef<HTMLDivElement, ThumbnailItemProps>(
...
                                 icon: <IconArrowCircleUp20 />,
-                                onClick: onOpenFileDialog,
+                                onClick: openFileDialog,
                             },
                             {
-                                title: 'Replace with asset',
+                                title: image ? 'Replace with asset' : 'Choose image',
```
Comment (ragi96) [inline packages/thumbnail-grid-block/src/components/Thumbnail/Item.tsx:118]:
```suggestion
                                title: image ? 'Replace with asset' : 'Browse asset',
```

To match the naming on the blockInjection

**Review by ragi96 — APPROVED** [review]
LGTM

**Pull request merged**